### ISSUE TITLE
Severe frame rate loss in proximity to Natura redwood trees v1.2.1

### ISSUE THREAD

**Issue by ghost** [Ambiotic/Ambiotic#33]
As the title says, when the player is in proximity to (approx. 8 blocks away from) redwood trees from the mod Natura, there is severe frame rate loss. This frame rate loss is made even worse with more redwoods close by; i.e. standing between 3 of them turns my game into a slideshow.

I haven't tested whether this occurs with large trees from other mods (e.g. sacred oaks from Biomes o'Plenty, Twilight Oaks from Twilight Forest, etc.), or whether it's just Natura redwood trees, but that may be worth investigating also.

**Comment by jimcarreer** [comment]
It may be more to do with the canopy of these trees causing the Player.IsExposed variable to struggle. This is actually a known problem to me: the IsExposed variable is powerful but also hard to determine efficiently.  I will try to investigate this as time permits, strangely though I have not had this problem with the regular Roofed Forest in Minecraft.  Could you possibly test the other forest Biomes you mentioned and let me know the results?

**Comment by ghost** [comment]
Sure, I'll test them tonight after work. Note that what I mentioned are trees, not biomes, so it's only the individual trees that cause the problem, not the biome the player is in. Another gigantic tree I'll test is the Sacred Rubber Tree from Minefactory Reloaded.

**Comment by jimcarreer** [comment]
As always you can get a hold of me on #minecraftforge.  I can probably provide you with a resource pack that will confirm my suspicions about what part of the code is causing the lag.  If you're already using the resource pack, you can extract engine.json and change **line 32** from:

```
"Type" : "Exposed",
```

to

```
"Type": "Skip",
```

This will disable the Player.IsExposed variable.  If you no longer have lag under the trees I imagine then that this variable was the cause of it. (This will break a bunch of stuff since I use that variable a lot in the emitters, I just want to know if it's the source of the lag).

I'll be back online around 6 EST, but I'll be idling in #minecraftforge.

**Comment by ghost** [comment]
Further testing has revealed that the problem does not occur with Robust Twilight Oak trees from the Twilight Forest mod. An interesting difference between the redwoods and the twilight oak is that the twilight oak has branches and leaf blocks at intervals starting fairly close to the ground, but the redwood doesn't have any leaves until the top of the tree. Another difference is that the Twilight Oak is hollow on the inside, but the redwood is solid all the way through.

The Biomes o'Plenty Sacred Oak also causes frame rate loss, despite being a lot shorter than the redwood.

MFR's Sacred and Mega Rubber Trees also cause frame rate loss similar to the Natura redwood trees.

As you suspected, once I turned off Player.IsExposed, following your guidelines, there was no framerate loss.

**Comment by jimcarreer** [comment]
Yup. So this is related to #26 in a way, the search algorithm I use is efficient in like 92% of cases but that other 8% are things like your tree, which make it work a little harder.  It's interesting that the Twilight Oak trees don't cause it but I suspect that may just be a lucky coincidence.

**Comment by jimcarreer** [comment]
@Tetrajak If you'd be willing to test it, I have a preliminary build that may resolve the issue.  Just let me know and Ill throw the link up so you can download the new version.

**Comment by ghost** [comment]
Absolutely, I'd be keen to test it. I can do so once I get home from work.

**Comment by jimcarreer** [comment]
https://www.dropbox.com/s/kx5mp7o92ohq5dk/ambiotic-test-1.2.2.jar?dl=0

Try that jar out for me see how it goes. There might be some lag still but I'm hoping its greatly reduced.

**Comment by ghost** [comment]
It crashed as soon as I got in range of the usual frame loss effects. I've uploaded the error log here; http://pastebin.com/4XaBG19c

**Comment by jimcarreer** [comment]
Whoops, forgot that list can be empty o.O. Do you want to jump on #minecraftforge and we can work this out real quick?

**Comment by DarianLStephens** [comment]
I don't know if you got any further with this test, but I've been having frame loss when in sheltered, semi-exposed areas, with losses as high as 58-59 frames from a steady 60.
http://imgur.com/a/jbgpF
My guess is that the redwood just provides a perfect environment for this issue to show itself, and is in not way directly related to the tree.
